File: python/examples/streaming/with_tools.py
import math

from mirascope import llm


@llm.tool
def sqrt_tool(number: float) -> float:
    """Computes the square root of a number."""
    return math.sqrt(number)


@llm.call("openai/gpt-5-mini", tools=[sqrt_tool])
def math_assistant(query: str):
    return query


response = math_assistant.stream("What's the square root of 4242?")

while True:  # Loop to ensure we execute all tool calls
    for chunk in response.pretty_stream():
        print(chunk, end="", flush=True)
    print()

    if not response.tool_calls:
        break

    tool_outputs = response.execute_tools()
    response = response.resume(tool_outputs)

File: typescript/examples/streaming/with-tools.ts
/**
 * Streaming with tools example.
 *
 * Demonstrates streaming responses while using tools.
 *
 * Run with: bun run example examples/streaming/with-tools.ts
 */
import { llm } from "mirascope";
import { z } from "zod";

const sqrtTool = llm.defineTool({
  name: "sqrt_tool",
  description: "Computes the square root of a number",
  validator: z.object({
    number: z.number().describe("The number to compute the square root of"),
  }),
  tool: ({ number }) => Math.sqrt(number),
});

const mathAssistant = llm.defineCall<{ query: string }>({
  model: "openai/gpt-4o-mini",
  tools: [sqrtTool],
  template: ({ query }) => query,
});

let response = await mathAssistant.stream({
  query: "What's the square root of 4242?",
});

// Loop to ensure we execute all tool calls
while (true) {
  // Stream the response
  for await (const text of response.textStream()) {
    process.stdout.write(text);
  }
  console.log();

  // Check if there are tool calls after consuming the stream
  if (response.toolCalls.length === 0) {
    break;
  }

  // Execute tools and resume
  const toolOutputs = await response.executeTools();
  response = await response.resume(toolOutputs);
}
